### PR TITLE
[RL] Preserve per-group KV cache across policy updates

### PR DESCRIPTION
**Pull request by wwwjn**
## Summary

- pin each rollout group to the generator policy version installed when the group enters the active window
- pass that version through both routing layers to vLLM as `cache_salt`, so new-policy groups cannot reuse old-policy KV while in-flight groups keep their reusable prefixes
- default to preserving KV across weight syncs, matching Prime-RL; add `generator.reset_kv_cache_on_weight_sync=true` as the opt-in cache-reset path
- keep batch-invariant recipes on the re-prefill path for exact trainer/generator comparisons

## Why

A policy update changes the model weights but does not automatically invalidate vLLM's cached KV. Letting a newly started rollout group hit KV created by an older policy would give it stale prefix state. Clearing the entire cache prevents that reuse, but it also preempts every active request and makes long-running, multi-turn groups prefill their accumulated context again. This change supports two explicit KV-cache lifecycle modes.

### Mode 1: preserve version-salted KV (default)

With `generator.reset_kv_cache_on_weight_sync=false`:

1. When a rollout group enters the active window, it pins the current policy version as `policy_version_at_start`. That value is passed as `cache_salt` for every request in the group.
2. A weight pull runs between `engine.step()` bursts. No request is admitted or advanced while the new weights are being installed, but existing requests and their KV blocks remain in vLLM.
3. After the pull, in-flight groups resume from their existing KV instead of being preempted and re-prefilled.
4. Groups admitted after the pull receive the new policy-version salt. They cannot hit old-salt blocks, so their prefixes are computed with the new weights.
5. Old-salt blocks remain reusable by their original groups until vLLM evicts them normally.

This mode avoids reset-and-reprefill work and preserves cache locality across policy updates. Its tradeoff is that the salt is only a prefix-hash namespace; it does not guarantee that every KV block used by an in-flight group came from one weight version. A group can retain prefix KV produced by its starting policy and append new KV after newer weights have been installed.

### Mode 2: reset and re-prefill

With `generator.reset_kv_cache_on_weight_sync=true`:

1. The generator installs the new weights.
2. It clears the prefix cache and preempts running requests.
3. vLLM reschedules those requests and recomputes their accumulated context with the new weights before decode continues.

This mode provides strict recomputation under the latest installed policy and remains useful for exact trainer/generator comparisons. The additional prefill compute may also help overlap FSDP parameter all-gathers. The cost is extra prefill work, queueing, and latency, especially for long-running multi-turn groups.

The default mode follows Prime-RL's behavior: a live-policy group uses its `policy_version_at_start` as the cache salt, while frozen-model traffic remains unsalted. TorchTitan obtains these semantics without calling vLLM's asynchronous `pause_generation` API because weight pulls are serialized between `engine.step()` bursts.

## DAPO Math performance measurement

Compared the default version-salted KV-preservation path (`--generator.no-reset-kv-cache-on-weight-sync`) with cache reset (`--generator.reset-kv-cache-on-weight-sync`).

### Setup

- `rl_dapo_qwen3_4b_math_8k`, Qwen3-4B-Base BF16
- 4x GB300: TP2 trainer plus two TP1 generators
- 10 optimizer steps; 8 prompt groups/step; 16 completions/prompt
- 2,048-token prompt and 8,192-token response limits
- Same checkpoint, config, seeds, and hardware; only the flag differed
- Jobs: preserve `1459386`; reset/re-prefill `1459394`

Generator request E2E is vLLM `last_token_ts - queued_ts`, including queue, prefill, and decode. Results are means across steps 2-10, excluding the step before the first policy-update reset.

| Metric | Preserve/salted reuse | Reset/re-prefill | Preserve delta |
| --- | ---: | ---: | ---: |
| Generator request E2E mean | 11,035.2 ms | 10,945.9 ms | +0.8% |
| Inter-token latency mean | 11.506 ms | 11.527 ms | -0.2% |
| Queue latency mean | 0.994 ms | 13.238 ms | -92.5% |
| Decode latency mean | 11,016.3 ms | 10,913.9 ms | +0.9% |
| Cached prompt tokens/request | 135.5 | 138.7 | -2.4% |
| Aggregate generation throughput estimate | 9,037 tok/s | 8,740 tok/s | +3.4% |
| Reported prefix-cache hit rate mean | 85.33% | 83.27% | +2.06 pp |

Preserving version-salted KV improved estimated generation throughput by 3.4%, prefix-cache hit rate by 2.06 percentage points, and queue latency by 92.5%. Request E2E was effectively unchanged. Throughput samples were noisy, so this is directional rather than statistically conclusive; an eight-GPU multi-turn run is the next measurement.

### CODE REVIEW

File: torchtitan/rl/generator.py [hunk 1]
```diff
@@ -1194,22 +1184,7 @@ async def _engine_loop(self) -> None:
...
-                        )
-                        for request, engine_input in zip(
-                            local_requests, engine_inputs, strict=True
-                        ):
-                            self._engine.add_request(
-                                request_id=request.request_id,
-                                prompt=engine_input,
-                                params=self._build_sampling_params(request.sampling),
-                            )
+                        self._admit_requests(local_requests)
```
Comment (tianyu-l) [inline torchtitan/rl/generator.py:1187]: since it's only used once, prefer to not creating a standalone function

File: torchtitan/rl/generator.py [hunk 2]
```diff
@@ -1299,6 +1274,26 @@ def _build_sampling_params(self, sampling: SamplingConfig) -> SamplingParams:
...
 
+    def _admit_requests(self, requests: list[GenerationRequest]) -> None:
+        """Render queued requests and add them to the local vLLM engine."""
+        # render_cmpl is vLLM's input pipeline. Tokenization is a no-op for
+        # tokenized prompts, and the typed input carries cache_salt to vLLM.
+        engine_inputs = self._engine.renderer.render_cmpl(
+            [
+                {
+                    "prompt_token_ids": request.prompt_token_ids,
+                    "cache_salt": request.cache_salt,
```
Comment (tianyu-l) [inline torchtitan/rl/generator.py]: so "cache_salt" is a vllm term?

Even if vllm use this term, can we use something more readable

File: torchtitan/rl/controller.py [hunk 3]
```diff
@@ -405,16 +405,6 @@ def __post_init__(self):
...
 
-            if (
-                not self.generator_router.hot_swap
-                and not self.generator.reset_prefix_cache_on_weight_sync
-            ):
-                raise ValueError(
-                    "generator_router.hot_swap=False requires "
-                    "generator.reset_prefix_cache_on_weight_sync=True, else requests admitted after a "
-                    "pull reuse KV cached under the old weights."
-                )
```
Comment (tianyu-l) [inline torchtitan/rl/controller.py:416]: makes sense as we introduce salt

File: torchtitan/rl/components/work_buffer.py [hunk 4]
```diff
@@ -32,11 +32,14 @@ class _RolloutGroupWorkState(enum.Enum):
...
-    The input loop sets `group_id` + `sample`; the buffer owns `state` and `rollout_group`
-    (`init=False`, so the input loop can't set them).
+    The input loop sets `group_id`, `policy_version_at_start`, and `sample`; the
+    buffer owns `state` and `rollout_group` (`init=False`, so the input loop
+    can't set them).
     """
 
     group_id: int
+    policy_version_at_start: int
+    """Generator policy version pinned when this rollout group entered the buffer."""
```
Comment (tianyu-l) [inline torchtitan/rl/components/work_buffer.py:42]: to confirm, in multi-turn, all the turns are sharing the same salt

File: torchtitan/rl/controller.py [hunk 5]
```diff
@@ -966,6 +965,10 @@ async def _rollout_loop(
                 return
             try:
                 with sl.log_trace_span("rollout_group"):
+                    generate_fn = self._make_generate_fn(
```
Comment (tianyu-l) [inline torchtitan/rl/controller.py:968]: why do we make `cache_salt` part of the `generate_fn`, instead of passing it into `run_group_rollouts`?

I think we have this `generate_fn` abstraction for potentially swapping it with an http service. Having salt in the function definition sounds weird.

File: torchtitan/rl/generator.py [hunk 6]
```diff
@@ -1121,6 +1110,7 @@ async def generate(
                     prompt_token_ids=prompt_token_ids,
                     sampling=sampling,
                     routing_session_id=routing_session_id,
+                    cache_salt=cache_salt,
```
Comment (tianyu-l) [inline torchtitan/rl/generator.py:1113]: so "cache_salt" is a vllm term?

I think we should be more careful when considering when to do the first `cache_salt=policy_version_at_start` renaming throughout the call chain.
- Maybe here constructing `GenerationRequest` is the first time that we should use vllm-specific terms?
- Maybe we should wait until the final `admit_request` from vllm engine.

File: torchtitan/rl/controller.py [hunk 7]
```diff
@@ -797,6 +790,7 @@ async def run(self) -> None:
             generator_router=self.generator_router,
             group_buffer=self._group_buffer,
             num_prompts_per_train_step=async_loop.num_prompts_per_train_step,
+            initial_policy_version=self.start_step,
```
Comment (tianyu-l) [inline torchtitan/rl/controller.py:793]: do we really need weight sync to know the `initial_policy_version`?

File: torchtitan/rl/controller.py [hunk 8]
```diff
@@ -837,7 +827,11 @@ async def run(self) -> None:
 
         # data_input_loop
         data_input_task = asyncio.create_task(
-            self._data_input_loop(self._group_buffer), name="data_input"
+            self._data_input_loop(
+                self._group_buffer,
+                weight_sync=self._weight_sync,
```
Comment (tianyu-l) [inline torchtitan/rl/controller.py:832]: definitely don't send weight_sync manager to data input loop, at least you can send version??

File: torchtitan/rl/distributed/weight_sync.py [hunk 9]
```diff
@@ -119,13 +126,14 @@ async def _generator_pull_and_release_buffer_slots(
             start = time.perf_counter()
             await self._generator_router.pull_model_state_dict.call_one(version)
```
Comment (tianyu-l) [inline torchtitan/rl/distributed/weight_sync.py:127]: what if some generator finishes weight sync, but other haven't? You would use a version as salt which you haven't seen the weight from.

File: torchtitan/rl/controller.py [hunk 10]
```diff
@@ -935,6 +934,7 @@ async def _data_input_loop(self, group_buffer: RolloutGroupWorkBuffer) -> None:
             await group_buffer.add_work(
                 RolloutGroupWork(
                     group_id=group_index,
+                    policy_version_at_start=weight_sync.generator_policy_version,
```
Comment (tianyu-l) [inline torchtitan/rl/controller.py:937]: I'm confused. Taking a step back, why do we need generator policy version to be the salt?

Is the goal of the salt is to let ALL (multi-turn) requests from a rollout group to share the same salt? Then why don't we use rollout group id???

**Review by tianyu-l — CHANGES_REQUESTED** [review]

**Comment by tianyu-l** [comment]
@claude 
review this PR, and review my review

File: torchtitan/rl/controller.py (hunk 10 above)
Comment (wwwjn) [reply torchtitan/rl/controller.py:937]: The salt is a "namespace" of the kv cache. When a kv cache query comes in, the query will first check the "namespace" then the corresponding position. We only need to update the "namespace" when the weights are being updated. 

(Will update this comment with more details, will also update in the code to document the behavior)
